### PR TITLE
descs: add ByIDGetter and ByNameGetter interfaces

### PR DESCRIPTION
**Pull request by postamar**
These pave the way for deprecating the existing `Get*ByID` and `Get*ByName` `Collection` methods.

Informs #87753.

Release note: None

### CODE REVIEW

**Comment by postamar** [comment]
This is ready for a round of review. This doesn't functionally change much, the purpose of this PR is to make it possible to sort the frightful mess that the `tree.CommonLookupFlags` and `tree.ObjectLookupFlags` have become, which in turn should lead to https://github.com/cockroachdb/cockroach/issues/87753.

**Review by ajwerner — COMMENTED** [review]
I love the direction. It makes a lot of sense to me. We'll need to be careful to not trash performance when we adopt this. Leaving my high-level feedback now and will come back to look at details. 

> Reviewed 13 of 34 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93813#-:-NJgljGk0GFK8Ris__dr:bvdzf54)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/catalog/descs/getters.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93813#-NJgkKTXB4T2yiaJzHat:-NJgkKTXB4T2yiaJzHau:b2ccnsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3afba2a786397ce670e282fba453580d9ff6d739/pkg/sql/catalog/descs/getters.go#L33)):*
> ```Go
> 
> // ByIDGetter looks up immutable descriptors by ID.
> type ByIDGetter interface {
> ```

I'm modestly worries that this interface approach is going to end up hurting from a performance perspective, at least, eventually. I worry that we're going to end up allocating frequently. I do get the value in doing this. I wonder if ultimately the approach we'll want to take is to return the concrete struct which implements this to the callers and have all the various forms actually share the same underlying pointer. Consider:

```
type ByIDGetter getterBase
```

You can then implement `ByIDGetter` the interface and have all the fields of a `getterBase` without exposing any of its methods or what not. Then, to convert between the various implementations, you can cast the values without ever incurring the need to allocate the object on the heap. 

Keep an eye on the various benchmarks we have related to resolving descriptors as you make progress around these parts. This code ends up being rather performance critica. 
___
*[`pkg/sql/catalog/descs/getters.go` line 578 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93813#-NJgje-L8CkQ6H54JMN1:-NJgje-L8CkQ6H54JMN2:b-ded9m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3afba2a786397ce670e282fba453580d9ff6d739/pkg/sql/catalog/descs/getters.go#L578)):*
> ```Go
> // TODO(postamar): clean up inconsistencies, enforce sane defaults.
> type getterFlags struct {
> 	context      contextFlags
> ```

the code might read more cleanly if you embed the `contextFlags`

**Review by Xiang-Gu — COMMENTED** [review]
I saw that you categorize all the flags into: context-related, layer-related, and desc-specific. This is very great! My question is that it seems you also introduced `withoutStorage` and `withoutCommittedAdding` flags bc they don't exist in the present tree.CommonLookupFlags. Why do you want to have these two new flags?

**Comment by postamar** [comment]
>  I wonder if ultimately the approach we'll want to take is to return the concrete struct which implements this to the callers and have all the various forms actually share the same underlying pointer.

Sounds reasonable. When I started sketching out these interfaces I didn't have a clear picture in mind as to where they'd lead to. Now that that's clearer I don't feel strongly that they _should_ be interfaces at all by any means. It's not like there'll be alternative implementations anyway.

> Why do you want to have these two new flags?

They were, in fact, in `tree.CommonLookupFlags`, so they're not new. They're flags that should be internal to the Collection, for sure. This is a big reason why I believe the flags should be cleaned up.

**Comment by postamar** [comment]
Thanks for the suggestion to get rid of the interfaces, it's much better without them. I ran some benchmarks and the impact is now negligible, and I attribute it to the `txn` interface which is a placeholder for what's to come in https://github.com/cockroachdb/cockroach/pull/93218 . 

Assuming the CI isn't sad this should be ready for a thorough review.

File: pkg/sql/schema_resolver.go [hunk 1]
```diff
@@ -150,10 +150,47 @@ func (sr *schemaResolver) LookupObject(
...
 	}
 
-	prefix, objMeta, err = sr.descCollection.GetObjectByName(ctx, sr.txn, dbName, scName, obName, flags)
+	prefix, objMeta, err = sr.GetObjectByName(
+		ctx, dbName, scName, obName, flags,
+	)
 	return objMeta != nil, prefix, objMeta, err
 }
 
+// GetObjectByName returns an object descriptor by name.
```
Comment (ajwerner) [inline pkg/sql/schema_resolver.go:159]: I think this comment should say something more about the structure of the various arguments. What is allowed to be empty and what is not?

**Review by ajwerner — APPROVED** [review]
This pleases me very much. In an ideal future world, we'll remove the `Txn` arg from the `*descs.Collection` functions and then can encapsulate that thing as an interface (it's always a pointer on the heap anyway). I think if we did that, it'd go a very long way towards making the catalog access more composable. What is currently the getters would be the layer to import and not `descs`.

**Pull request merged**